README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# Ring Button Group

This package provides a configurable component to support a group of buttons
displayed on a sized ring.

<img src="https://github.com/xeric/ring_button_group/raw/main/images/demo.png" title="" alt="" width="561">

## Features

<img src="https://github.com/xeric/ring_button_group/raw/main/images/demo.gif" title="" alt="" width="561">

## Getting started

```shell
flutter pub add ring_button_group
```

or manually add this to your package's `pubspec.yaml` file:

```yaml
...
dependencies:
  ring_button_group: ^1.0.0
```

Now in your Dart code, you can use:

```dart
import 'package:ring_button_group/ring_button_group.dart';
```

## Usage

```dart
SizedBox(
    width: 200, //set a width
    height: 200, //set a height, should be same with width
    child: RingButtonGroup(
      buttonNumber: 5, //total number of buttons in group
      icons: const [ //a list of icons displayed on button
        Icon(Icons.abc, color: Colors.white,),
        Icon(Icons.baby_changing_station, color: Colors.white,),
        Icon(Icons.cabin, color: Colors.white,),
        Icon(Icons.dangerous, color: Colors.white,),
        Icon(Icons.e_mobiledata, color: Colors.white,),
      ],
      type: RingButtonGroupType.MULTIPLE_SELECTABLE, // allow multiple select
      pressedIndex: const {1}, //default selected buttons index, start from 0
      shadowEffect: true,
      onPressed: (index, allSelected) {  },
    ),
  ),
)
```

## RingButtonGroup Attributes

| Attribute          | Type               | Default                  | Annotation                                                                                                                                        |
| ------------------ | ------------------ | ------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| buttonNumber       | int                |                          | total number of buttons in this group, must be greater than 1                                                                                     |
| pressedIndex       | Set<int>           | `{}`                     | a list of indexes of button which should be pressed when rendering, index starts from 0                                                           |
| onPressed          | OnPressedFunction? | `null`                   | Function(int index, Set<int>? selected), selected is a set of indexes represents current all pressed buttons when type is **MULTIPLE_SELECTABLE** |
| disabled           | bool               | `false`                  | navigation buttons of **BottomBarWithSheet**                                                                                                      |
| buttonSize         | double             | `40`                     | the size of button, it identifies the radius of the circle                                                                                        |
| toneColor          | Color              | `Colors.blueAccent`      | main color of the button                                                                                                                          |
| tintColor          | Color              | `Colors.blueGrey`        | button color while pressing down                                                                                                                  |
| activeColor        | Color              | `Colors.lightBlueAccent` | button color after pressed                                                                                                                        |
| borderColor        | Color              | `Colors.lightBlueAccent` | button border color                                                                                                                               |
| disableColor       | Color              | `Colors.blueAccent`      | disabled main color, when color same with toneColor, auto transform to grayscale                                                                  |
| disableBorderColor | Color              | `Colors.lightBlueAccent` | disabled border color, when color same with borderColor, auto transform to grayscale                                                              |
| icons              | List<Icon>?        | `null`                   | button icons list, while set, the list length must be same with `buttonNumber`                                                                    |
| labels             | List<Text>?        | `null`                   | button label list, only works when `icons` is `null`, the list length must be same with `buttonNumber`                                            |
| child              | Widget?            | null                     | a child of ring button, common case is put a circle in the center for display purpose                                                             |
| shadowEffect       | bool               | false                    | use a inner shadow effects in pressed/selected button                                                                                             |

## RingButtonGroupType

| type                | description                                                                    |
| ------------------- | ------------------------------------------------------------------------------ |
| PRESS_ONLY          | such type only allow press action, but will not stay in pressed/selected state |
| SINGLE_SELECTABLE   | only allow one of buttons stay in pressed/selected state, like a radio box     |
| MULTIPLE_SELECTABLE | allow multiple pressed like a checkbox                                         |

## Example

to `/example` folder.

you can find the example show in screenshot from file: `./example/lib/main/dart`

#### Run example in your simulator or device

```shell
flutter run
```

## Additional information

when you using this ring button group, this widget should be contained by a `SizedBox` or similar widget with width and height properties, and must be a square - with same value in width and height, the widget itself is stretched.


CHANGELOG.md
## 1.0.0
* First formal version release

## 1.0.1
* fix README.md image url to absolute one to let pub.dev can display image properly.

test/ring_button_group_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:ring_button_group/ring_button_group.dart';

void main() {
  test('initialization', () {
    final calculator = RingButtonGroup(buttonNumber: 5, pressedIndex: const {1}, onPressed:(index, selected) {},);
    expect(calculator.buttonSize, 40);
  });
}


test/cyclic_button_group_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:cyclic_button_group/cyclic_button_group.dart';

void main() {
  test('adds one to input values', () {
    final calculator = CyclicButtonGroup(buttonNumber: 5, pressedIndex: const {1}, onPressed:(index, selected) {},);
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/ring_button_group.dart
library ring_button_group;

import 'package:flutter/material.dart';
import 'dart:math' as math;

class RingButtonGroup extends StatefulWidget {
  //how many buttons on ring
  final int buttonNumber;

  //list button index (start from 0) pressed at beginning, notice, when type is SINGLE_SELECTABLE, set size must less 2
  final Set<int> pressedIndex;

  //function when button pressing, args: index - index of pressing button, selected: all buttons now selected (only available when MULTIPLE_SELECTABLE)
  final OnPressedFunction? onPressed;

  //entire group disabled
  final bool disabled;

  //type of behavior - see: RingButtonGroupType
  final RingButtonGroupType type;

  //ring button size |<-size->|----o----|------|(bulleye)
  final double buttonSize;

  //main color of button
  final Color toneColor;

  //pressing color
  final Color tintColor;

  //pressed color
  final Color activeColor;

  //disabled border color, when color same with borderColor, auto transform to grayscale
  final Color disableBorderColor;

  //disabled main color, when color same with toneColor, auto transform to grayscale
  final Color disableColor;

  //border color of buttons
  final Color borderColor;

  //icons on button, size must be same with buttonNumber
  final List<Icon>? icons;

  //labels on button, size must be same with buttonNumber
  final List<Text>? labels;

  //child widget of ring button group
  final Widget? child;

  //allow shadow effects when button pressed
  final bool shadowEffect;

  const RingButtonGroup({
    super.key,
    required this.buttonNumber,
    this.pressedIndex = const {},
    this.onPressed,
    this.buttonSize = 40,
    this.disabled = false,
    this.type = RingButtonGroupType.SINGLE_SELECTABLE,
    this.toneColor = Colors.blueAccent,
    this.tintColor = Colors.blueGrey,
    this.activeColor = Colors.lightBlueAccent,
    this.disableBorderColor = Colors.lightBlueAccent,
    this.disableColor = Colors.blueAccent,
    this.borderColor = Colors.lightBlueAccent,
    this.icons,
    this.labels,
    this.child,
    this.shadowEffect = false,
  })  : assert(buttonNumber > 1),
        assert(labels != null ? labels.length == buttonNumber : true),
        assert(icons != null ? icons.length == buttonNumber : true),
        assert(type == RingButtonGroupType.SINGLE_SELECTABLE
            ? pressedIndex.length < 2
            : true);

  @override
  State<StatefulWidget> createState() => RingButtonGroupState();
}

class RingButtonGroupState extends State<RingButtonGroup> {
  late ButtonStatus status;

  @override
  void initState() {
    super.initState();
    status = ButtonStatus(widget.pressedIndex, false);
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (context, constraints) {
      var widgets = <Widget>[];
      widgets.addAll([for (var i = 0; i < widget.buttonNumber; i++) i]
          .map((index) => GestureDetector(
                onTapCancel: () {
                  if (!widget.disabled) {
                    setState(() {
                      status._pressedDown = null;
                    });
                  }
                },
                onTapUp: (detial) {
                  if (!widget.disabled) {
                    setState(() {
                      switch (widget.type) {
                        case RingButtonGroupType.SINGLE_SELECTABLE:
                          status.pressed.clear();
                          status.pressed.add(index);
                          break;
                        case RingButtonGroupType.MULTIPLE_SELECTABLE:
                          !status.pressed.remove(index)
                              ? status.pressed.add(index)
                              : null;
                          break;
                        default:
                      }

                      widget.onPressed!(
                          index,
                          widget.type == RingButtonGroupType.MULTIPLE_SELECTABLE
                              ? status._pressed
                              : null);

                      status._pressedDown = null;
                    });
                  }
                },
                onTapDown: (detial) {
                  setState(() {
                    status._pressedDown = index;
                  });
                },
                child: FractionallySizedBox(
                    widthFactor: 1,
                    heightFactor: 1,
                    child: Stack(
                      children: [
                        FractionallySizedBox(
                          widthFactor: 1,
                          heightFactor: 1,
                          child: CustomPaint(
                              painter: RingButtonPainter(
                                  toneColor: widget.toneColor,
                                  tintColor: widget.tintColor,
                                  activeColor: widget.activeColor,
                                  disableColor: widget.disableColor,
                                  disableBorderColor: widget.disableBorderColor,
                                  borderColor: widget.borderColor,
                                  buttonNumber: widget.buttonNumber,
                                  buttonSize: widget.buttonSize,
                                  buttonIndex: index,
                                  pressed: status.pressed.contains(index),
                                  disabled: widget.disabled,
                                  shadowEffect: widget.shadowEffect,
                                  pressDown: status._pressedDown == index)),
                        ),
                        RingButtonIcon(
                          index: index,
                          buttonSize: widget.buttonSize,
                          size: constraints.minWidth,
                          total: widget.buttonNumber,
                          child: FractionalTranslation(
                              translation: const Offset(-0.5, -0.5),
                              child: widget.icons != null
                                  ? widget.icons![index]
                                  : widget.labels != null
                                      ? widget.labels![index]
                                      : null),
                        ),
                      ],
                    )),
              )));
      widget.child != null ? widgets.add(widget.child!) : null;
      return Stack(children: widgets);
    });
  }
}

class RingButtonIcon extends Positioned {
  final int index;
  final int total;

  final double size;
  final double buttonSize;
  //double right;

  RingButtonIcon(
      {super.key,
      required this.index,
      required this.total,
      required this.buttonSize,
      required this.size,
      required super.child})
      : super(
          left: calcLeft(total, index, buttonSize, size),
          top: calcTop(total, index, buttonSize, size),
        );

  static double calcLeft(int total, int index, double buttonSize, double size) {
    double angle01 = 360 / total * index + 360 / total / 2;
    return math.sin(math.pi * angle01 / 180) * (size / 2 - buttonSize / 2) +
        size / 2;
  }

  static double calcTop(int total, int index, double buttonSize, double size) {
    double angle01 = 360 / total * index + 360 / total / 2;
    return size / 2 -
        math.cos(math.pi * (360 - angle01) / 180) * (size / 2 - buttonSize / 2);
  }
}

class RingButtonPainter extends CustomPainter {
  final Color toneColor;
  final Color tintColor;
  final Color activeColor;
  final Color disableBorderColor;
  final Color disableColor;
  final Color borderColor;
  final int buttonNumber;
  final int buttonIndex;
  final double buttonSize;
  final bool pressed;
  final bool pressDown;
  final bool disabled;
  final bool shadowEffect;

  Path path = Path();

  RingButtonPainter(
      {required this.toneColor,
      required this.buttonNumber,
      required this.buttonSize,
      required this.borderColor,
      required this.buttonIndex,
      required this.pressed,
      required this.pressDown,
      required this.tintColor,
      required this.activeColor,
      required this.disableBorderColor,
      required this.disableColor,
      required this.disabled,
      required this.shadowEffect});

  @override
  void paint(Canvas canvas, Size size) {
    var paintStroke = Paint()
      ..color = disabled
          ? disableBorderColor == borderColor
              ? getGreyScale(borderColor)
              : disableBorderColor
          : borderColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1.0;
    var paintFill = Paint()
      ..color = disabled
          ? disableColor == toneColor
              ? getGreyScale(toneColor)
              : disableColor
          : pressDown && !pressed
              ? tintColor
              : (pressed
                  ? shadowEffect
                      ? toneColor
                      : activeColor
                  : toneColor)
      ..style = PaintingStyle.fill
      ..strokeWidth = 4.0;
    // paintFill.maskFilter = pressDown ? MaskFilter.blur(BlurStyle.normal, convertRadiusToSigma(5)) : paintFill.maskFilter;
    var shadowPaint = shadowEffect
        ? (Paint()
          ..color = activeColor.withAlpha(245)
          ..maskFilter = MaskFilter.blur(
              BlurStyle.inner, convertRadiusToSigma(buttonSize / 5)))
        : null;

    double innerRadius = size.width / 2 - buttonSize;
    double outterRadius = size.width / 2;

    _drawButton(
        canvas,
        paintStroke,
        paintFill,
        !disabled && pressed ? shadowPaint : null,
        path,
        innerRadius,
        outterRadius,
        _calcOffsets(buttonIndex, innerRadius, outterRadius, size.width / 2));
  }

  static double convertRadiusToSigma(double radius) {
    return radius * 0.57735 + 0.1;
  }

  List<Offset> _calcOffsets(
      int index, double innerRadius, double outterRadius, double base) {
    double angle01 = 360 / buttonNumber * index;
    // print(angle01);
    double p0x = math.sin(math.pi * angle01 / 180) * innerRadius + base;
    double p0y = base - math.cos(math.pi * (360 - angle01) / 180) * innerRadius;

    double p1x = math.sin(math.pi * angle01 / 180) * outterRadius + base;
    double p1y =
        base - math.cos(math.pi * (360 - angle01) / 180) * outterRadius;

    double angle23 = 360 / buttonNumber * (index + 1);
    double p3x = math.sin(math.pi * angle23 / 180) * innerRadius + base;
    double p3y = base - math.cos(math.pi * (360 - angle23) / 180) * innerRadius;

    double p2x = math.sin(math.pi * angle23 / 180) * outterRadius + base;
    double p2y =
        base - math.cos(math.pi * (360 - angle23) / 180) * outterRadius;

    return [
      Offset(p0x, p0y),
      Offset(p1x, p1y),
      Offset(p2x, p2y),
      Offset(p3x, p3y)
    ];
  }

  void _drawButton(
      Canvas canvas,
      Paint paintStroke,
      Paint paintFill,
      Paint? shadowPaint,
      Path path,
      double innerRadius,
      double outterRadius,
      List<Offset> offsets) {
    path.moveTo(offsets[0].dx, offsets[0].dy);
    path.lineTo(offsets[1].dx, offsets[1].dy);
    path.arcToPoint(offsets[2], radius: Radius.circular(outterRadius));
    path.lineTo(offsets[3].dx, offsets[3].dy);
    path.arcToPoint(offsets[0],
        radius: Radius.circular(innerRadius), clockwise: false);
    path.close();

    canvas.drawPath(path, paintFill);
    canvas.drawPath(path, paintStroke);
    shadowPaint != null ? canvas.drawPath(path, shadowPaint) : null;
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }

  @override
  bool hitTest(Offset position) {
    return path.contains(position);
  }
}

class ButtonStatus {
  final Set<int> _pressed = <int>{};
  int? _pressedDown;
  bool _disabled = false;

  Set<int> get pressed => _pressed;

  bool get isDisabled => _disabled;

  int? get pressedDown => _pressedDown;

  ButtonStatus(Set<int> pressed, bool disabled) {
    _pressed.addAll(pressed);
    _disabled = disabled;
  }
}

typedef OnPressedFunction = Function(int index, Set<int>? selected);

enum RingButtonGroupType { PRESS_ONLY, SINGLE_SELECTABLE, MULTIPLE_SELECTABLE }

Color getGreyScale(Color orginalColor) {
  int gray = (orginalColor.red * 0.199).toInt() +
      (orginalColor.green * 0.387).toInt() +
      (orginalColor.blue * 0.414).toInt();
  return Color.fromARGB(orginalColor.alpha, gray, gray, gray);
}


example/lib/main.dart
import 'package:ring_button_group/ring_button_group.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Ring Button Group'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late String displayName = "Test";

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Center(
              // Center is a layout widget. It takes a single child and positions it
              // in the middle of the parent.
              child: SizedBox(
                width: 200,
                height: 200,
                child: RingButtonGroup(
                  buttonNumber: 5,
                  icons: const [
                    Icon(Icons.abc, color: Colors.white,),
                    Icon(Icons.baby_changing_station, color: Colors.white,),
                    Icon(Icons.cabin, color: Colors.white,),
                    Icon(Icons.dangerous, color: Colors.white,),
                    Icon(Icons.e_mobiledata, color: Colors.white,),
                  ],
                  type: RingButtonGroupType.MULTIPLE_SELECTABLE,
                  pressedIndex: const {1},
                  shadowEffect: true,
                  onPressed: (index, selected) {  },
                ),
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Center(
              // Center is a layout widget. It takes a single child and positions it
              // in the middle of the parent.
              child: SizedBox(
                width: 200,
                height: 200,
                child: RingButtonGroup(
                  disabled: true,
                  buttonNumber: 6,
                  buttonSize: 60,
                  labels: const [
                    Text("I"),
                    Text("II"),
                    Text("III"),
                    Text("IV"),
                    Text("V"),
                    Text("VI"),
                  ],
                  type: RingButtonGroupType.SINGLE_SELECTABLE,
                  pressedIndex: const {1},
                  onPressed: (index, selected) {  },
                ),
              ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Center(
              // Center is a layout widget. It takes a single child and positions it
              // in the middle of the parent.
              child: SizedBox(
                width: 200,
                height: 200,
                child: RingButtonGroup(
                  buttonNumber: 6,
                  buttonSize: 50,
                  toneColor: Colors.deepOrange,
                  borderColor: Colors.orange,
                  activeColor: Colors.deepOrange.shade200,
                  tintColor: Colors.deepOrange.shade400,
                  labels: const [
                    Text("1", style: TextStyle(color: Colors.white),),
                    Text("2", style: TextStyle(color: Colors.white),),
                    Text("3", style: TextStyle(color: Colors.white),),
                    Text("4", style: TextStyle(color: Colors.white),),
                    Text("5", style: TextStyle(color: Colors.white),),
                    Text("6", style: TextStyle(color: Colors.white),),
                  ],
                  type: RingButtonGroupType.SINGLE_SELECTABLE,
                  // pressedIndex: {},
                  onPressed: (index, selected) {
                    setState(() {
                      displayName = "Test $index";
                    });
                  },
                  child: Center(
                    child: Container(
                      decoration: BoxDecoration(
                        border: Border.all(color: Colors.orange),
                        borderRadius: BorderRadius.circular(40),
                        boxShadow: const [BoxShadow(blurStyle: BlurStyle.outer, color: Colors.black12, blurRadius: 5)],
                      ),
                      width: 80,
                      height: 80,
                      child: Center(child: Text(displayName))
                      ),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
